### PR TITLE
SQLAlchemy: Add support for 'autogenerate_uuid' in column creation

### PR DESCRIPTION
**Pull request by surister**
Solves #533 

Where by setting the column:
```py
Column('x', sa.Text, primary_key=True, crate_autogenerate_uuid=True)
```

will generate:
```SQL
x STRING DEFAULT gen_random_text_uuid() NOT NULL
```

## Checklist

 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] Docs
 - [x] Tests

### CODE REVIEW

**Comment by surister** [comment]
Circling back to this, I'm effectively checking against
```py
Column(..., server_default='value')
```
with

```python
if default is not None:
    raise sa.exc.CompileError(
            "Can only have one default value but server_default"
            " and crate_generate_uuid are set"
               )
```

But It'd probably also clash with `Column(..., default=lambda:str(uuid.uuid4()))`

I will have a deeper look at this.

File: src/crate/client/sqlalchemy/tests/create_table_test.py [hunk 1]
```diff
@@ -311,3 +311,36 @@
              'pk STRING NOT NULL, \n\t'
              'answer INT DEFAULT 42, \n\t'
              'PRIMARY KEY (pk)\n)\n\n'), ())
+
+    def test_column_autogenerate_uuid(self):
+        class DummyTable(self.Base):
```
[the remark this answers is not in the document]
Comment (amotl) [reply src/crate/client/sqlalchemy/tests/create_table_test.py:316]: Please forget about those, they are just false positives.

- https://github.com/github/codeql/issues/11407

[the remark this answers is not in the document]
Comment (amotl) [reply src/crate/client/sqlalchemy/tests/create_table_test.py:316]: @aibaars just mentioned at https://github.com/github/codeql/issues/11427#issuecomment-1721059096:

> The [advanced-security/dismiss-alerts](https://github.com/advanced-security/dismiss-alerts) Action implements support for inline suppression comments and automatically dismisses alerts in GitHub Code Scanning.

We may want to give it a shot, depending on mood and time.

[the remark this answers is not in the document]
Comment (amotl) [reply src/crate/client/sqlalchemy/tests/create_table_test.py:316]: I am not sure. Using `# codeql[py/unused-local-variable]` might work already, or not.

File: src/crate/client/sqlalchemy/tests/create_table_test.py [hunk 2]
```diff
@@ -311,3 +311,36 @@ class DummyTable(self.Base):
...
+
+        with self.assertRaises(sa.exc.CompileError):
+            self.Base.metadata.create_all(bind=self.engine)
+
+    def test_column_autogenerate_uuid_clashes_with_server_default(self):
+        class DummyTable(self.Base):
+            __tablename__ = 't'
+            pk = sa.Column(sa.String, primary_key=True,
+                           crate_autogenerate_uuid=True,
+                           server_default='value')
```
Comment (amotl) [inline src/crate/client/sqlalchemy/tests/create_table_test.py:342]: Ah, using both `crate_autogenerate_uuid` and `server_default` at the same time. Yeah, that may want to produce a failure, no? What do you think, @seut?

**Comment by surister** [comment]
> Hi @surister,
> 
> thanks for your patch. It looks good to me so far.
> 
> Regarding your question about option clashes: I don't know if and how SQLAlchemy prevents to use both `default` and `server_default`, or if it doesn't do at all.
> 
> Depending on how SQLAlchemy handles that, we should not necessarily add anything on top - only if it makes sense, and doesn't overcomplicate the code.
> 
> With kind regards, Andreas.

Using `default` + `crate_autogenerate_uuid` doesn't really 'clash', it seems that CrateDB just ignores the `DEFAULT` directive and puts in whatever value it gets on the `INSERT`

File: src/crate/client/sqlalchemy/tests/create_table_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aibaars) [reply src/crate/client/sqlalchemy/tests/create_table_test.py:316]: Yes they should work, and plain `#noqa` should work too. Just make sure to run [AlertSuppression.ql](https://github.com/github/codeql/blob/a3e2e420f016e569a4edcdf5fc9703584192357f/python/ql/src/AlertSuppression.ql) in addition to the normal CodeQL rules.

File: docs/sqlalchemy.rst [hunk 3]
```diff
@@ -207,6 +207,7 @@ system <sa:orm_declarative_mapping>`:
     ...     quote_ft = sa.Column(sa.String)
     ...     even_more_details = sa.Column(sa.String, crate_columnstore=False)
     ...     created_at = sa.Column(sa.DateTime, server_default=sa.func.now())
+    ...     uuid = sa.Column(sa.String, crate_autogenerate_uuid=True)
```
Comment (mfussenegger) [inline docs/sqlalchemy.rst:210]: Doesn't `id = sa.Column(sa.String, primary_key=True, server_default=func.gen_random_text_uuid())` already work? Why do we need a custom property?

**Comment by surister** [comment]
More importantly when using it as:
```py
Column(..., crate_autogenerate_uuid=True, primary_key=True)
```

We get the following warning:

```
SAWarning: Column 'aaaa3.x' is marked as a member of the primary key for table 'aaaa3', but has no Python-side or server-side default generator indicated, nor does it indicate 'autoincrement=True' or 'nullable=True', and no explicit value is passed.  Primary key columns typically may not store NULL.
```

Because technically, there is no autoincrement_column and `server_default` nor `default` is set, so SQLAlchemy cannot assure that the value will not be `Null` (and nullable in this case is set to False by default)

[SQLAlchemy warning source](https://github.com/agronholm/sqlalchemy-1/blob/d76cb7213557c24609a1a75d8c391aea0179562a/lib/sqlalchemy/sql/crud.py#L559)

File: docs/sqlalchemy.rst (hunk 3 above)
Comment (amotl) [reply docs/sqlalchemy.rst:210]: You are having a valid point here. Maybe the `crate_autogenerate_uuid` should be a table- or dialect-level parameter instead, to cover situations like outlined in GH-575, which is a slightly different topic, and can be deferred to another patch.

If the user can write their own model definitions, your proposal to use `server_default=func.gen_random_text_uuid()` indeed would be completely satisfactory.

So, the outcome for this topic could be essentially to cover that detail about `func.gen_random_text_uuid()` within the documentation and software tests at
- docs/sqlalchemy.rst
- src/crate/client/sqlalchemy/tests/create_table_test.py

and not add any specific functionality for this on the column level. Apologies that GH-533 was giving wrong guidance here.

**Comment by amotl** [comment]
Dear @surister,

> So, the outcome for this topic could be essentially to cover that detail about `func.gen_random_text_uuid()` within the documentation at, for example, `docs/sqlalchemy.rst`.

I've just submitted GH-585, so I am closing this. Apologies again that I was giving wrong guidance on this matter through GH-533.

With kind regards,
Andreas.

**Pull request closed without merging**